sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -1407,3 +1440,12 @@ case class UpCast(child: Expression, dataType: DataType, walkedTypePath: Seq[Str
   extends UnaryExpression with Unevaluable {
   override lazy val resolved = false
 }
+
+/**
+ * Cast the child expression to the target data type, but will throw error if the cast violates
+ * the store assignment rules of ANSI SQL, e.g. string -> int, array -> string.
+ */
+case class AssignableCast(child: Expression, dataType: DataType, walkedTypePath: Seq[String] = Nil)
```

Comment:
can we apply it for the v2 table insertion first? We probably need another PR to apply it for v1 table insertion, as it's a behavior change and needs migration guide.

Reply:
BTW I'm not sure if we need `walkedTypePath`. `Upcast` needs it because it needs to record the field path in a class. I think for `AssignableCast` we only need to record the column name.